feat(masthead): adding shadow parts

bruno-amorim opened the pull request:
### Related Ticket(s)

[ADCMS-5170](https://jsw.ibm.com/browse/ADCMS-5170)

### Description

Adding shadow parts to the Masthead component so it can be styles outside of the Shadow Tree.
Snapshots updated.

jkaeser commented on packages/web-components/src/components/masthead/left-nav-menu-section.ts:
```diff
@@ -222,16 +226,20 @@ class C4DLeftNavMenuSection extends HostListenerMixin(FocusMixin(LitElement)) {
         ${showBackBtn
           ? html`
               <li
+                part="masthead-side-nav-submemu"
```
I see that you're referencing a CSS class to come up with the name here, but in this case, I think there are a couple issues with that:
1. We don't need to copy over typos
2. This isn't really functioning as a submenu. Rather, it's functioning as a menu item and a back button wrapper:
```suggestion
                part="menu-item back-button-wrapper"
```

jkaeser commented on packages/web-components/src/components/masthead/left-nav-menu-section.ts:
```diff
@@ -222,16 +226,20 @@ class C4DLeftNavMenuSection extends HostListenerMixin(FocusMixin(LitElement)) {
                 class="${prefix}--side-nav__menu-item ${prefix}--masthead__side-nav--submemu-back"
                 role="none">
                 <button
+                  part="side-nav-link"
```
```suggestion
                  part="menu-link back-button"
```

jkaeser commented on packages/web-components/src/components/masthead/left-nav-menu-section.ts:
```diff
@@ -222,16 +226,20 @@ class C4DLeftNavMenuSection extends HostListenerMixin(FocusMixin(LitElement)) {
                   @click="${handleClickBack}">
-                  <span class="${prefix}--side-nav__link-text"
+                  <span
+                    part="side-nav-link-text"
```
```suggestion
                    part="menu-link-text back-button-text"
```

jkaeser reviewed (changes requested):
Remember, given the way these will be used we don't need to prefix each part name with the component name. E.g.
```css
/* Redundant */
c4d-left-nav-menu-category-heading::part(left-nav-menu-category-h2) { ... }

/* Not redundant */
c4d-left-nav-menu-category-heading::part(heading) { ... }
```

I'd also encourage you to take a look at the markup as rendered in the browser to get a better idea of how all these pieces are actually used. While looking at class names for hints is good, they don't always necessarily match how the element is actually used. See my comment on `<c4d-left-nav-menu-section>`'s back button for an example of what I mean.

It looks like all of these part names can be simplified. I left some notes for how I might go about doing that for some components. Can you do another round trying to match that approach? Thanks!

bruno-amorim commented:
Hey guys @andy-blum @Valentin-Sorin-Nicolae @marcelojcs @m4olivei 

I have addressed the comments from John, can you please check it now? I have just simplified the naming convention for all the shadow parts.

Thank you!!

m4olivei commented on packages/web-components/src/components/masthead/left-nav-menu-section.ts:
```diff
@@ -222,16 +226,20 @@ class C4DLeftNavMenuSection extends HostListenerMixin(FocusMixin(LitElement)) {
                 class="${prefix}--side-nav__menu-item ${prefix}--masthead__side-nav--submemu-back"
                 role="none">
                 <button
+                  part="menu-link back-button"
```
The `back-button` part needs documentation.

m4olivei commented on packages/web-components/src/components/masthead/masthead-global-bar.ts:
```diff
@@ -61,7 +62,11 @@ class C4DMastheadGlobalBar extends FocusMixin(
     });
-    return html` <div class="${classes}"><slot></slot></div> `;
+    return html`
+      <div part="ce-header-global-container" class="${classes}">
```
I think we can simplify this to just `container`.
```suggestion
      <div part="container" class="${classes}">
```

m4olivei commented on packages/web-components/src/components/masthead/masthead-l1.ts:
```diff
@@ -999,7 +1085,7 @@ class C4DMastheadL1 extends StableSelectorMixin(LitElement) {
 
   static get dropDownToggleEvent() {
-    return `${c4dPrefix}-masthead-l1-dropdown-toggle`;
+    return `${c4dPrefix}-l1-dropdown-toggle`;
```
Is this change intentional? Does not seem releated to the shadow parts 🤔

m4olivei commented on packages/web-components/src/components/masthead/masthead-menu-button.ts:
```diff
@@ -115,6 +117,7 @@ class C4DMastheadMenuButton extends HostListenerMixin(CDSHeaderMenuButton) {
   private _renderSentinel = (side: String) => {
     return html`
       <button
+        part="sentinel-button"
```
Maybe add a modifier part so we can target the specific button if needed.
```suggestion
        part="sentinel-button sentinel-button--${side}"
```

m4olivei commented on packages/web-components/src/components/masthead/megamenu-category-link.ts:
```diff
@@ -59,13 +59,13 @@ class C4DMegaMenuCategoryLink extends CDSLink {
     return html`
       ${title
         ? html`
-            <div part="link-heading">
```
I'd rather leave this as is, in case it's already being used in the wild. This is an API break.

m4olivei commented on packages/web-components/src/components/masthead/megamenu-category-link.ts:
```diff
@@ -59,13 +59,13 @@ class C4DMegaMenuCategoryLink extends CDSLink {
             </div>
           `
         : ''}
-      <span part="link-description">
```
Same here. Let's leave as it.

m4olivei commented on packages/web-components/src/components/masthead/top-nav.ts:
```diff
@@ -575,7 +592,9 @@ class C4DTopNav extends StableSelectorMixin(HostListenerMixin(CDSHeaderNav)) {
-                      <div class="${prefix}--sub-content-left"></div>
+                      <div
+                        part="sub-content-left"
+                        class="${prefix}--sub-content-left"></div>
```
Missed a `<div>` element in this file on line 601: `<div class="${caretLeftContainerClasses}">`

m4olivei commented on packages/web-components/src/components/masthead/top-nav.ts:
```diff
@@ -601,10 +622,16 @@ class C4DTopNav extends StableSelectorMixin(HostListenerMixin(CDSHeaderNav)) {
+                  class="${c4dPrefix}-ce--header__nav-content-container">
+                  <div
+                    part="header-nav-content"
+                    class="${prefix}--header__nav-content">
```
Missed a `<div>` element on line 639: `<div class="${prefix}--sub-content-left"></div>`

m4olivei reviewed (changes requested):
Few things found. Overall nice work. Almost there!

m4olivei commented on packages/web-components/src/components/masthead/masthead-menu-button.ts:
```diff
@@ -25,6 +25,10 @@ const { prefix, stablePrefix: c4dPrefix } = settings;
  * @element c4d-masthead-menu-button
+ * @csspart sentinel-button - The sentinel button for managing focus.
+ * @csspart ce-header-menu-trigger-container - Container for the menu trigger button.
+ * @csspart sentinel-button-start - The start sentinel button for managing focus. Usage: `c4d-masthead-menu-button::part(sentinel-button-start)`
```
```suggestion
 * @csspart sentinel-button--start - The start sentinel button for managing focus. Usage: `c4d-masthead-menu-button::part(sentinel-button--start)`
```

m4olivei commented on packages/web-components/src/components/masthead/masthead-menu-button.ts:
```diff
@@ -25,6 +25,10 @@ const { prefix, stablePrefix: c4dPrefix } = settings;
+ * @csspart sentinel-button - The sentinel button for managing focus.
+ * @csspart ce-header-menu-trigger-container - Container for the menu trigger button.
+ * @csspart sentinel-button-start - The start sentinel button for managing focus. Usage: `c4d-masthead-menu-button::part(sentinel-button-start)`
+ * @csspart sentinel-button-end - The end sentinel button for managing focus. Usage: `c4d-masthead-menu-button::part(sentinel-button-end)`
```
```suggestion
 * @csspart sentinel-button--end - The end sentinel button for managing focus. Usage: `c4d-masthead-menu-button::part(sentinel-button--end)`
```

m4olivei commented on packages/web-components/src/components/masthead/masthead-menu-button.ts:
```diff
@@ -115,6 +119,7 @@ class C4DMastheadMenuButton extends HostListenerMixin(CDSHeaderMenuButton) {
   private _renderSentinel = (side: String) => {
     return html`
       <button
+        part="sentinel-button sentinel-button-${side}"
```
```suggestion
        part="sentinel-button sentinel-button--${side}"
```

m4olivei reviewed (changes requested):
Small nits. Will commit.

m4olivei commented on packages/web-components/src/components/masthead/masthead-l1.ts:
```diff
@@ -524,6 +575,7 @@ class C4DMastheadL1 extends StableSelectorMixin(LitElement) {
                   return html`
                     <li>
```
This `<li>` is missing a part attribute. Maybe `l1-dropdown-menu-item`.

m4olivei commented on packages/web-components/src/components/masthead/masthead-l1.ts:
```diff
@@ -447,20 +486,25 @@ class C4DMastheadL1 extends StableSelectorMixin(LitElement) {
-            ? html`<div class="${prefix}--masthead__l1-dropdown-announcement">
+            ? html`<div
+                part="l1-dropdown-announcement"
+                class="${prefix}--masthead__l1-dropdown-announcement">
```
Not able to comment directly on it, but line 511 below is missing a `part` attribute. Probably `dropdown-links` would work.

m4olivei commented on packages/web-components/src/components/masthead/masthead-l1.ts:
```diff
@@ -656,7 +730,7 @@ class C4DMastheadL1 extends StableSelectorMixin(LitElement) {
           ${footer
             ? html`
-                <a
+                <a part="l1-dropdown-viewall"
```
This is good. Found another missing `part` on line 946. The `<a>` around an L1 column heading. I would suggest `part="l1-dropdown-item-link l1-dropdown-item-link--heading"`

m4olivei reviewed (changes requested):
Sorry found a couple more.
